Add per-role timeouts

psarna opened the pull request:
This series allows adding per-role parameters. The current list of parameters includes:
 * `read_timeout` - custom timeout for read operations
 * `write_timeout` - custom timeout for write operations

In order to set up per-role parameters, one should use the already existing CQL API for roles:
```cql
CREATE ROLE example WITH options = {'read_timeout': 50ms}
```
or
```cql
ALTER ROLE example WITH options = {'read_timeout': 1s, 'write_timeout': 500ms}
```

Once a session with given role is established, it will use per-role timeouts instead of globally configured
timeouts, if there are any.

Role options can be viewed with the standard `LIST ROLES` statement:
```cql
LIST ROLES;

 role      | super | login | options
-----------+-------+-------+-----------------------------------------------------
 cassandra |  True |  True | {'read_timeout': '50ms', 'write_timeout': '1000us'}

```

When a timeout configuration is present in multiple roles, the most strict combination of them is effective for a user.
So if a user is a member of role `cassandra` with config `{'read_timeout': '50ms', 'write_timeout': '20ms'}`, another role `read_intensive` with config `{'read_timeout': 10ms}` and role `cassandra_parent` is granted to role `cassandra` itself, and has config `{'write_timeout': 100ms}`, then the effective configuration for this user is `{'read_timeout': 10ms, 'write_timeout': '20ms'}`, since only then the timeout restrictions for all roles are satisfied.

NOTE: timeouts are currently propagated only within the local node. Propagation to other nodes is planned for a follow-up.

Tests: unit(dev), manual(cqlsh)

avikivity commented:
/cc @haaawk

haaawk commented:
It seems that it will be supported by drivers automatically by just sending the "ALTER SESSION" cql, right?

kostja commented on cql3/statements/alter_session_statement.cc:
```diff
@@ -0,0 +1,94 @@
+future<::shared_ptr<cql_transport::messages::result_message>>
+alter_session_statement::execute(service::storage_proxy& proxy, service::query_state& state, const query_options& options) const {
+    state.get_client_state().set_session_params(props->get_params());
+    auto result = ::make_shared<cql_transport::messages::result_message::void_message>();
```
How is the driver going to find out that session properties were altered? Will it have to parse CQL to find out? 
I though you were going to return a metadata flag indicating that session properties were changed.

How do I retreive session properties using CQL?

haaawk replied:
It's kept in client state which as far as I remember has a CQL table that stores the info for all client states.

psarna replied:
That's how I imagined it, given that parsing this statement is a bunch of string splits. We could even go one step further and represent it as JSON for free parsing.

As for retrieving the info - I don't have a good idea. How does SQL do that, with a local (virtual) table?

psarna replied:
Oh, right, we have `system.clients` now, which is a very good fit.

kostja replied:
I think the drivers should not have to parse CQL to track session changes.

psarna replied:
I generally agree, but modifying the binary protocol is also a mess, since it requires the driver and the server to agree on who understands what. In case of the driver side, it could actually be made even simpler (and require no parsing), if the session altering is never sent in a raw form, but rather built by the driver itself, e.g. from a native map implementation (something like that in pseudopython: `session.execute(AlterSessionStatement({'foo': 'bar', 'foo2': 'baz'}))`).
That's how `USE <keyspace>` works in gocql, as far as I understood from my consultation with PiotrD - sending raw `USE` statements is disallowed, and the only way to use a keyspace is to set it up via gocql API.

avikivity replied:
Ah. But what about cqlsh?

psarna replied:
cqlsh parses some statements:
```python
        if custom_handler:
            parsed = cqlruleset.cql_whole_parse_tokens(tokens, srcstr=srcstr,
                                                       startsymbol='cqlshCommand')
```
and then
```python
    def do_use(self, parsed):
        ksname = parsed.get_binding('ksname')
        success, _ = self.perform_simple_statement(SimpleStatement(parsed.extract_orig()))
        if success:
            if ksname[0] == '"' and ksname[-1] == '"':
                self.current_keyspace = self.cql_unprotect_name(ksname)
            else:
                self.current_keyspace = ksname.lower()
```

avikivity replied:
> I generally agree, but modifying the binary protocol is also a mess, since it requires the driver and the server to agree on who understands what. In case of the driver side, it could actually be made even simpler (and require no parsing), if the session altering is never sent in a raw form, but rather built by the driver itself, e.g. from a native map implementation (something like that in pseudopython: `session.execute(AlterSessionStatement({'foo': 'bar', 'foo2': 'baz'}))`).
> That's how `USE <keyspace>` works in gocql, as far as I understood from my consultation with PiotrD - sending raw `USE` statements is disallowed, and the only way to use a keyspace is to set it up via gocql API.

I agree. Another option, which I think is inferior, is to send ALTER SESSION and have the server respond with a flag that says "this needs to go to all connections". The driver sees the flag and broadcasts the alter.

psarna commented:
> It seems that it will be supported by drivers automatically by just sending the "ALTER SESSION" cql, right?

It will, but it needs special treatment, just like `USE <keyspace>` does. When a new connection is established, such stateful information needs to be sent to the new connection. E.g. in gocql, when a new connection is established and a keyspace is configured, `USE <keyspace>` is automatically sent by the driver.

haaawk commented:
> > It seems that it will be supported by drivers automatically by just sending the "ALTER SESSION" cql, right?
> 
> It will, but it needs special treatment, just like `USE <keyspace>` does. When a new connection is established, such stateful information needs to be sent to the new connection. E.g. in gocql, when a new connection is established and a keyspace is configured, `USE <keyspace>` is automatically sent by the driver.

Ok I opened https://github.com/scylladb/java-driver/issues/48

avikivity commented on cql3/statements/alter_session_prop_defs.cc:
```diff
@@ -0,0 +1,71 @@
+#include "concrete_types.hh"
+
+void cql3::statements::alter_session_prop_defs::validate() {
+    static const std::set<sstring> keywords({"latency_limit_for_reads", "latency_limit_for_writes"});
```
We might (or not) want to recommend that if you have different latency goals for reads and writes, you should use different sessions.

avikivity replied:
(if we do, then there's be just latency, not latency for read and latency for write).

psarna replied:
Yup, may make sense. I split it up in this RFC just because reads and writes do usually have totally different latencies in metrics, but it's all arbitrary for now. Users probably have a single latency requirement and don't care if the underlying operation is a read, a write or something in-between.

psarna replied:
Actually, there was one more reason I split the values - scylla config has separate timeouts for these two. Is it Cassandra legacy which we can ignore, or do users perhaps sometimes set up different configs for these two?

avikivity commented on cql3/statements/alter_session_prop_defs.cc:
```diff
@@ -0,0 +1,71 @@
+ *
+ * Modified by ScyllaDB
+ */
+
```
There is no need to dual-license new code, just code that has Apache heritage. In addition, it turns out that the copyright statement is sufficient to satisfy the license terms and there is no need for "Modified by ScyllaDB".

avikivity commented:
Looks good. How will the driver know to broadcast the information? Parsing CQL seems like a bad choice.

psarna commented:
Um, was it closed by mistake?

avikivity commented:
Yes :(

avikivity commented:
Someone is remote controlling my mouse, there is no other explanation.

psarna commented:
Maybe the battery is finally running out.

avikivity commented:
> Maybe the battery is finally running out.

It was replaced a while ago after it died with zero warning. So it's probably some other battery dying.

psarna commented:
As for the drivers knowing whether they should broadcast the information, here are the 3 summarized options:
* parse the CQL and recognize the information
  * precedent: cqlsh does that with `USE <keyspace>`
* require the driver to broadcast on its own, e.g. by forbidding raw `ALTER SESSION` statements and forcing the user to use driver-side primitives, which know how to broadcast to all existing and future connections of a session
  * precedent: gocql does that with `USE <keyspace>`
* get the information back via metadata
  * unprecedented, requires binary protocol changes

I'm in favor of option 2 with optional 1. Adding more metadata is an overkill for me, since this CQL is trivial to parse client-side with a regular expression, and we don't even have to do that if we require the driver to forbid raw queries and provide internal broadcasting mechanisms

avikivity commented:
I agree with your 2.

The binary protocol has some support for custom payloads, perhaps we can fit it as a custom payload during the initial connection handshake. In fact we can use OPTIONS/STARTUP.

psarna commented:
I pushed v2, which is still RFC due to lack of unit tests, but it has:
 * a way of deleting set params
 * latency limits which are actually enforced -- i.e. the queries fail if they don't manage to finish in time
 * a docs entry in docs/cql-extensions.md, which contain a short writeup with examples
 * a way of checking current values - via `system.clients` table

and it was manually tested in order to see if latency limits work for reads/writes.

avikivity commented on docs/cql-extensions.md:
````diff
@@ -41,4 +41,38 @@ The option can be specified at `CREATE TABLE` or `ALTER TABLE` queries in the sa
+ALTER SESSION
+	DELETE latency_limit_for_reads
+```
+
````
(Focusing review here)

Naming bikeshed: `read_timeout` and `write_timeout`. We don't actually limit the latency, we time out,
Proposal: add `timeout` (or `latency_limit`) that takes effect if for_read/for_write isn't specified. This way you can set a general latency limit with one value.

We should also have a way to specify that a session is not for realtime ops, that is, it's okay to send empty pages as keepalives instead of abandoning a slow query.

psarna replied:
Yeah, I even started calling these variables "timeouts" in docs and commit messages, since that's what they actually are. I'll rename them all to timeouts.

psarna replied:
I also considered adding `timeout` or `default_timeout`, but right now `read_timeout` adds a single branch to the select path, and `write_timeout` adds a single branch to the modification path. Having a default per-session parameter would add yet another branch, while not giving any huge benefit to the users - I think they are equally fine with setting `session.set_read_timeout(50ms).set_write_timeout(50ms)` as they would be with `session.set_timeout(50ms)`. Perhaps this idea can be revisited if we later expand the parameters to have tens of different timeouts, in which case it will definitely be helpful to add a default. It can be done later as a follow-up with very little effort, since the whole machinery would already be in place.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -266,6 +266,12 @@ uint64_t select_statement::do_get_limit(const query_options& options, ::shared_p
+db::timeout_clock::duration select_statement::get_timeout(const service::client_state& state, const query_options& options) const {
+    return state.get_session_params().latency_limit_for_reads.value_or(
+        options.get_timeout_config().*get_timeout_config_selector()
+    );
```
Suggest to populate the session timeout from the global timeout if not specified. This way, the runtime path is if-less.

avikivity replied:
(in fact we can just populate query_options::_timeout_config, then no changes are needed here)

psarna replied:
That's a great idea, we already have the information in place when query_options are created. Will do.

psarna replied:
Actually, `query_options` only keeps a reference to the global server config, which can't simply be modified. So it's a question of a branch in the code vs passing the whole timeout config for each query options. I think that the timeout config is heavy enough to avoid copying it extensively (7 `std::duration`s now, can grow).

avikivity replied:
We can make it refer to the session instead of the global config (and seed the session config from the global config).

psarna replied:
Yeah, but then we'll also need to verify if these timeouts are reloadable online from config. If they are, we'd need to propagate changes to new sessions too, which makes it a little more complicated. But maybe these configs are only reloadable upon restart, I haven't checked.

psarna replied:
*existing sessions, not new ones

avikivity replied:
Even if we add reload, we can make it re-propagate the timeouts to sessions with no timeout overrides.

psarna replied:
Right, that's even better. A single copy of the timeout config will be added for each connection, but in return, the timeout config reference for each executed statement will not be needed anymore, so it's usually a net win for memory usage (unless somebody uses a new connection for each statement, in which case they're doomed anyway).

avikivity replied:
In fact we're all doomed.

psarna commented:
@kostja pointed out that this syntax isn't exactly based on standard sql, so ill rework it a little bit. I think I'll base it on ALTER SESSION SET x = y from here: https://docs.oracle.com/cd/B19306_01/server.102/b14200/statements_2012.htm . It only takes a single parameter, which makes the implementation considerably easier while it still has same capabilities (it would just require more requests to be sent for multiple params, but it's a very rare operation anyway).

psarna commented:
v3:
 - extended the documentation and design decisions in docs/cql-extensions.md
 - added unit tests 
 - simplified the `ALTER SESSION SET` syntax in order to make it look like valid SQL, deleting params is performed simply by typing `SET x = NULL`
 - renamed all mentions of `latency_limit` to `timeout`

not done:
 - `default_timeout` parameter: left for later, if the list of params is expanded enough for it to be needed
 - populating the timeout_config struct in query_options directly: turns out that this object is passed by reference, probably due to its rather large size, so for now I decided to stay with adding a single additional branch only in affected paths - select statement and modification statements

psarna commented:
Also, I'll rework this issue's description in a moment to reflect all the changes

avikivity commented:
What about a driver-friendly version?

We can use custom payloads

```
    0x04: Custom payload flag. For a request or response frame, this indicates
          that a generic key-value custom payload for a custom QueryHandler
          implementation is present in the frame. Such a custom payload is simply
          ignored by the default QueryHandler implementation.
          Currently, only QUERY, PREPARE, EXECUTE and BATCH requests support
          payload.
          Type of custom payload is [bytes map] (see below).
```

They are under-specified but workable. Or we can have the driver construct ALTER SESSION statements and send them.

psarna commented:
v4:
 - moved timeout config from being a reference in query_options to being a full object inside client_state. That way, we don't need any branches when extracting timeouts for various statements, and ALTER SESSION can simply change the values in that object.
 
**NOTICE**: The last commit of this series is quite massive, although 95% of its contents is trivial. It consists mostly of removing a `timeout` parameter (in millions of call sites), which is no longer needed after timeouts are moved to client_state.

I suggest reviewing the last patch separately.

psarna commented:
> What about a driver-friendly version?
> 
> We can use custom payloads
> 
> ```
>     0x04: Custom payload flag. For a request or response frame, this indicates
>           that a generic key-value custom payload for a custom QueryHandler
>           implementation is present in the frame. Such a custom payload is simply
>           ignored by the default QueryHandler implementation.
>           Currently, only QUERY, PREPARE, EXECUTE and BATCH requests support
>           payload.
>           Type of custom payload is [bytes map] (see below).
> ```
> 
> They are under-specified but workable. Or we can have the driver construct ALTER SESSION statements and send them.

I assume you mean adding a custom payload in the response to ALTER SESSION (or some other rare query, like PREPARE), which contains the serialized map of options, so that drivers don't have to bother figuring it out from the statement itself? That's a good idea, and the presence of such a map could also let the driver know that this map needs to be propagated to all connections - so we'll essentially have a flag returned by the server as Kostja suggested and drivers won't have to parse anything or expose special interfaces, but at the same time we will circumvent all the backward compatibility pains of changing the binary protocol.

That said, I'd humbly suggest to leave this part as a follow-up - if I understand correctly, it's independent of the other changes, and Scylla currently doesn't even acknowledge that there's a custom payload flag:
```cpp
enum cql_frame_flags {
    compression = 0x01,
    tracing     = 0x02,
    warning     = 0x08,
};
```
, so adding it here would be a first step towards making it an ever-growing-never-fully-mergeable series.

psarna commented:
(rebased to get rid of the conflicts, no other changes)

avikivity commented:
> > What about a driver-friendly version?
> > We can use custom payloads
> > ```
> >     0x04: Custom payload flag. For a request or response frame, this indicates
> >           that a generic key-value custom payload for a custom QueryHandler
> >           implementation is present in the frame. Such a custom payload is simply
> >           ignored by the default QueryHandler implementation.
> >           Currently, only QUERY, PREPARE, EXECUTE and BATCH requests support
> >           payload.
> >           Type of custom payload is [bytes map] (see below).
> > ```
> > 
> > 
> > They are under-specified but workable. Or we can have the driver construct ALTER SESSION statements and send them.
> 
> I assume you mean adding a custom payload in the response to ALTER SESSION (or some other rare query, like PREPARE), which contains the serialized map of options, so that drivers don't have to bother figuring it out from the statement itself? 

I did not do so far as to form an idea.

> That's a good idea,

Thanks!

> and the presence of such a map could also let the driver know that this map needs to be propagated to all connections - so we'll essentially have a flag returned by the server as Kostja suggested and drivers won't have to parse anything or expose special interfaces, but at the same time we will circumvent all the backward compatibility pains of changing the binary protocol.
> 
> That said, I'd humbly suggest to leave this part as a follow-up - if I understand correctly, it's independent of the other changes, and Scylla currently doesn't even acknowledge that there's a custom payload flag:
> 
> ```c++
> enum cql_frame_flags {
>     compression = 0x01,
>     tracing     = 0x02,
>     warning     = 0x08,
> };
> ```
> 
> , so adding it here would be a first step towards making it an ever-growing-never-fully-mergeable series.

Certainly. We just need to be sure we see there is a path.

haaawk commented:
I've got one question. Why do we need new CQL statement when we could connect those settings with users like we do with workload prioritization? It seems that setting up expectations like online/batch and timeout length could fit well with current workload prioritization parameter connected to the user, no? @eliransin what do you think?

psarna commented:
As for users, does creating one require elevated privileges? If so, then managing timeouts via users will be less dynamic than a CQL statement. But perhaps it's not an issue, I don't know how often clients would want their timeouts changed. Also, if we expect a single client to potentially have many different timeout requirements, it could result in explosion of users, since one would need to be created for each config. These concerns does not disqualify this idea at all, but I just dump them here for more context. I'm also a total noob wrt. user-based workload proritization so I hope that somebody more experienced will state their mind.

haaawk commented:
I agree that connecting the parameters with users has some limitations but on the other hand it's much simpler and at least in my opinion much less error prone. New CQL statement has to be tested in multiple scenarios like appearing in a batch or in a prepared statement. Or being executed concurrently with different parameters. I understand if it turns out to be necessary but if not then I would suggest piggy-backing on the existing user mechanism.

psarna commented:
@avikivity what do you think about piggy-backing on roles? I'll spend today and tomorrow on verifying how it could work, but since I have access to `authenticated_user` from `client_state`, it would be quite easy to extract timeout params from a role when establishing a session. I'll still need a statement which adds role-specific params instead of session-specific one, but this one would require zero cooperation from the drivers, which sounds like a big advantage. We also lose a little bit of flexibility if somebody wants to quickly change their timeout, but maybe it's a non-issue.

Most of the code from this series will remain usable, I'll just add a notion of per-role params and add a statement to alter that, and then, when a session is created, I'll populate the per-session timeouts with whatever is stored in roles. There will also be some more questions wrt. how to compute timeouts if somebody has 5 roles and each one has a different timeout, but such dillemas are also solvable with `std::min`.

psarna commented:
(we already have `ALTER ROLE`, so I could just expand that with some timeout params and this oltp/olap boolean)

dekimir commented:
What happens if a user has multiple roles with conflicting timeouts assigned to them?

psarna commented:
I mentioned that in the last paragraph of https://github.com/scylladb/scylla/pull/7617#issuecomment-734299082 - I don't know yet, but one reasonable option would be `std::min`. I think it's still safe to assume that users would use separate, mutually exclusive roles created specifically for timeouts, but even if it's not the case, we can fall back to picking most strict timeouts possible.

psarna commented:
Turns out that the `options` map is not really supported by any of our authenticators, but that should be quite easy to add. And after that's done, I'll rework this pull request to be based on roles and present it for review. The main advantage would be that it involves no chances in CQL grammar whatsoever. The changes might take a while though, so I'll probably send something next week.

psarna commented:
v5: pivot!
 * instead of `ALTER SESSION`, timeouts are now set per role, via already existing `ALTER ROLE` statement
 * most of the commits actually stays the same
 * `options` parameter map is implemented for PasswordAuthenticator - it was already designed in the API, but it wasn't implemented
 * sessions now consult per-role timeouts right after being established (and authenticated)

avikivity commented:
Using roles make a lot of sense. We do that in enterprise too for workload prioritization (but with an extra indirection level).

I still like ALTER SESSION so that you can experiment with cqlsh.

avikivity commented on service/client_state.hh:
```diff
@@ -180,10 +194,16 @@ public:
             , _is_internal(true)
             , _is_thrift(false)
+            , _default_timeout_config(infinite_timeout_config)
+            , _timeout_config(infinite_timeout_config)
```
Aye, this is also simplifying code later on.

avikivity commented on db/system_distributed_keyspace.cc:
```diff
@@ -155,17 +155,20 @@ future<> system_distributed_keyspace::stop() {
 }
 
-static const timeout_config internal_distributed_timeout_config = [] {
+static service::query_state& internal_distributed_query_state() {
```
I'd feel happier if there was a new query_state for each internal query, but that's outside the scope here. We're making all internal queries behave as if they were one large query.

psarna replied:
I'll put it on the follow-up list, since I'd prefer to not complicate this patch, it's already a little messy.

avikivity commented on auth/password_authenticator.cc:
```diff
@@ -66,6 +66,7 @@ constexpr std::string_view password_authenticator_name("org.apache.cassandra.aut
 
 // name of the hash column.
 static constexpr std::string_view SALTED_HASH = "salted_hash";
+static constexpr std::string_view OPTIONS = "options";
```
I don't understand why this is related to the authenticator. Authenticator options should be about how the authenticator works, and we shouldn't need per-authenticator code for something generic like session configuration.

There are two options (hah!) here:
1. CREATE ROLE WITH OPTIONS is not used by any authenticator, and we find something to convince us that OPTIONS is not related to the authenticator. In this case we should redirect OPTIONS from the authenticator to the role manager.
2. CREATE ROLE WITH OPTIONS _is_ used by some authenticator or other, and so we create or reuse a keyword for out map. For example CREATE ROLE WITH SESSION OPTIONS = { ... }.

@dekimir any wisdom about roles here? Can we store stuff in the roles table even if the user uses a non-default role manager?

psarna replied:
Ok, I got more familiar with role manager, and I pinpointed the source of confusion. The design of our code really heavily suggests that it's the authenticator that stores custom **per-role** options. Here's a snippet from `list_roles_statement::execute`:
```cpp
    return do_for_each(sorted_roles, [&results, &rm, &a](const sstring& role) {
        return when_all_succeed(
                rm.can_login(role),
                rm.is_superuser(role),
                a.query_custom_options(role)).then_unpack([&results, &role](
                ...
```
For each role, we extract some information from the role manager, but then query the authenticator (`a`) for custom per-role options.

So, if I follow the review correctly, I should instead extend the role manager to use the already existing `options` table of `system_auth.roles` to store per-role options, and then the password authenticator could get options support for free, since the table it uses would already be populated.

I think that the main source of my confusion is that `system_auth.roles` is used both for general management of roles, regardless of which authenticator is used, and for storing PasswordAuthenticator-specific data, like passwords.

Also, the current status of authenticators supporting their custom options is that none of them do. And from what I saw in enterprise sasl authenticator, it claims supporting additional options, but there are no actual options implemented, so it basically just supports options by always returning an empty option set if asked.

@avikivity I now know what caused my confusion, but I still don't really have a clear idea on how to proceed. Should I rework `ALTER ROLE x WITH options = ` to fill our table instead of going straight to authenticators, which do not implement custom options? The current infrastructure is quite heavily based on the fact that these options belong to authenticators (e.g. the underlying authenticators all validate if they support options, and refuse to alter roles if they don't)

avikivity commented on cql3/statements/role-management-statements.cc:
```diff
@@ -59,6 +59,7 @@
 #include "gms/feature_service.hh"
 #include "transport/messages/result_message.hh"
 #include "unimplemented.hh"
+#include "concrete_types.hh"
```
(commenting on 391d1f2b210d5b157d63fee951d1d67fc068cb98 changelog)

I think we do need to update running sessions. Perhaps it can be in a follow-up, but we should follow-up quickly and make this easy to use instead of requiring a rolling restart.

avikivity replied:
(and probably iterating over all connections is fine)

avikivity commented on service/client_state.cc:
```diff
@@ -53,13 +53,64 @@
+                _default_timeout_config.read_timeout : read_timeout;
+        _timeout_config.range_read_timeout = read_timeout == db::timeout_clock::duration::max() ?
+                _default_timeout_config.range_read_timeout : read_timeout;
+         _timeout_config.write_timeout = write_timeout == db::timeout_clock::duration::max() ?
```
bad indent detected

psarna replied:
bad indent eliminated

avikivity commented on docs/roles.md:
```diff
@@ -0,0 +1,33 @@
+
+## Supported authenticators
+
+Roles are currently supported in PasswordAuthenticator.
```
As I mentioned before, roles and authenticators are supposed to be orthogonal. The challenge here is if we will be able to use the table-based role manager storage with non-standard role managers. @dekimir

psarna replied:
I see that we already do use this table - but only for accessing `is_superuser` and `can_login` columns. `salted_hash` is used only for PasswordAuthenticator and `options` was left totally untouched until this series.

avikivity commented:
I like this a lot. I think there is some confusion re. password_authenticator, but the direction is good.

Looking forward to adding non-OLTP sessions support.

but: we'll want ALTER SESSION for cqlsh.

psarna commented:
> I like this a lot. I think there is some confusion re. password_authenticator, but the direction is good.
I think that it makes sense for the authenticator to be responsible for per-role options. The whole CQL machinery looks designed specifically for that purpose:
 * authenticators declare if they support per-role options or not
 * if they do, they store it in any way they see fit:
   - password authenticator uses a system table
   - a hypothetical LDAP authenticator uses its internal LDAP tables, stored somewhere else

Timeouts are perhaps not a great example, but if we also stored other types of options (e.g. adding some custom permissions for specific roles), then the authenticator *should* be responsible for keeping these safe.

> 
> Looking forward to adding non-OLTP sessions support.
That would be scheduled for follow-up № 2, right after follow-up № 1, which will be all about refreshing existing sessions.
> 
> but: we'll want ALTER SESSION for cqlsh.

Sure. The patches can be simply revived from previous branches, they're almost free to restore.

avikivity commented:
> > I like this a lot. I think there is some confusion re. password_authenticator, but the direction is good.
> 
> I think that it makes sense for the authenticator to be responsible for per-role options. The whole CQL machinery looks designed specifically for that purpose:
> 
> * authenticators declare if they support per-role options or not
> * if they do, they store it in any way they see fit:
>   
>   * password authenticator uses a system table
>   * a hypothetical LDAP authenticator uses its internal LDAP tables, stored somewhere else
> 

But role information is stored using role_manager, not authenticator.

Say we have a kerberos authenticator, or OAuth2 (so you can log in via your gmail account). Where will the timeouts be stored?

> Timeouts are perhaps not a great example, but if we also stored other types of options (e.g. adding some custom permissions for specific roles), then the authenticator _should_ be responsible for keeping these safe.

That's done be the role manager today.

psarna commented:
On Sun, Nov 29, 2020, 19:02 Avi Kivity <notifications@github.com> wrote:

> I like this a lot. I think there is some confusion re.
> password_authenticator, but the direction is good.
>
> I think that it makes sense for the authenticator to be responsible for
> per-role options. The whole CQL machinery looks designed specifically for
> that purpose:
>
>    -
>
>    authenticators declare if they support per-role options or not
>    -
>
>    if they do, they store it in any way they see fit:
>    - password authenticator uses a system table
>       - a hypothetical LDAP authenticator uses its internal LDAP tables,
>       stored somewhere else
>
> But role information is stored using role_manager, not authenticator.
>
> Say we have a kerberos authenticator, or OAuth2 (so you can log in via
> your gmail account). Where will the timeouts be stored?
>
I assumed that some authenticators simply don't support options and it's
expected.


Timeouts are perhaps not a great example, but if we also stored other types
> of options (e.g. adding some custom permissions for specific roles), then
> the authenticator *should* be responsible for keeping these safe.
>
> That's done be the role manager today.
>
Alright, I have to admit that I have no clue what role manager is. I'll
read about it tomorrow and see how I can try to integrate it with custom
options. Meanwhile, maybe I'll just pivot back to ALTER SESSION if we want
it anyway, and do the roles later... I'll educate myself more and decide
tomorrow.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/7617#issuecomment-735432351>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACPTEFYVK2INHZXTW3I7HF3SSKEB7ANCNFSM4TXAE2AQ>
> .
>

avikivity commented:
Need input from @dekimir

psarna commented:
Btw, Cassandra docs say that the options field is:
```
OPTIONS
    Reserved for external authenticator plug-ins.
```

psarna commented:
I think I'll use the time difference between here and Canada to speculatively work on moving timeout-specific options from the authenticator to `system_auth.roles`. Then, even if @dekimir's input proves that it's a bad idea, then I'll at least be more skilled at role management.

psarna commented:
Wait, what I was even talking about... **I** added the `options` column in this very series. So, since all authenticators already use some columns from `system_auth.roles`, then it's definitely fine to simply add one more. I'll rework this series and send role-manager-based options soon. I simply got confused because the `salted_hash` column is only used by password authenticator, and I assumed that this whole table is only used by this authenticator, which is not true.

avikivity commented:
I think there's a good chance this is correct. Things like `is_superuser` and `can_login` are unlikely to be available in ldap schemas, so they have to be satisfied internally.

Indeed ldap_role_manager asks the default role manager to manage these attributes.

psarna commented:
v6:
 - moved per-role options to role manager from password authenticator,
   which makes them available for each supported authenticator

Current state of the follow-up list:
 - add propagating timeouts to all live sessions
 - add oltp/olap parameter
 - revive ALTER SESSION from previous iteration and make it available (for the sake of cqlsh and making it easier to test different values without having to have elevated privileges)

dekimir commented on service/client_state.hh:
```diff
@@ -315,6 +328,7 @@ public:
                                       auth::command_desc::type = auth::command_desc::type::OTHER) const;
     future<> has_schema_access(const schema& s, auth::permission p) const;
 
+    future<> update_per_role_params();
```
Please add a blurb specifying how timeouts are inherited and conflicts among them resolved.  Also explain why this must be done for all roles rather than a subset of them.

psarna replied:
done

dekimir commented on test/boost/auth_test.cc:
```diff
@@ -216,3 +217,49 @@ SEASTAR_TEST_CASE(alter_opts_on_system_auth_tables) {
+
+        // Only valid timeout values are accepted
+        BOOST_REQUIRE_THROW(e.execute_cql("ALTER ROLE user WITH options = {'read_timeout': 'I am not a valid duration'}").get(), marshal_exception);
+        BOOST_REQUIRE_THROW(e.execute_cql("ALTER ROLE user WITH options = {'read_timeout': 5us}").get(), exceptions::invalid_request_exception);
```
What about an over-24h case?

psarna replied:
Done. Also, fixed a regression, because it was possible to sneak up days or months by specifying a large enough number of seconds/hours - cql_duration parsing works differently than I initially expected.

dekimir commented on test/boost/auth_test.cc:
```diff
@@ -216,3 +217,49 @@ SEASTAR_TEST_CASE(alter_opts_on_system_auth_tables) {
+        auth::authentication_options opts {
+            .password = "pass"
+        };
+        auth::create_role(e.local_auth_service(), "user", config, opts).get();
```
What about cases of role inheritance, including multiple inheritance with conflicting timeouts?

psarna replied:
Done, also fixed a regression - timeouts weren't updated after `GRANT ROLE`.

dekimir commented:
My two cents: role options should be handled by role_manager, since authenticator doesn't deal with non-login roles.

psarna commented on db/query_context.hh:
```diff
@@ -55,10 +55,18 @@ struct query_context {
 
-        return do_with(timeout_config{d, d, d, d, d, d, d}, [this, req = std::move(req), &args...] (auto& tcfg) {
+        struct timeout_context {
+            std::unique_ptr<service::client_state> client_state;
```
The context instance gets moved, and one of its attributes (`query_state`) takes reference to another one of its attributes (`client_state`). In order for this reference to remain valid, I have to ensure that `client_state` has a stable address - and the simplest solution is to put it on the heap. Not very nice from the performance perspective, but I also don't expect it to show on perf as top offender anywhere.

avikivity replied:
This "not the top offender" ensures we can add allocations everywhere, since none of them will be a top offender.

But the untyped_result_set variants and not used in the data plane, so okay.

psarna replied:
@avikivity but note that this `execute_cql_with_timeout` helper function was introduced for paxos - which means that these extra allocation happens on LWT path. My gut tells me that it's still worth it to make internal calls use their own client state and that this allocation will not make a noticeable impact, but I haven't measured the actual overhead.
This allocation of a new client state was avoided so far, because paxos effectively used a global client state with specific timeout config, but it's no longer possible after timeout configuration is moved to the client state itself.

psarna replied:
On the other hand, if we really care about optimizing this path, raw CQL strings should just be dropped altogether, the whole cql layer should be bypassed for paxos and the request should go straight to storage proxy, like we do in alternator. So this single allocation is not the root cause of potential slowness.

dekimir replied:
> one of its attributes (query_state) takes reference to another one of its attributes (client_state)

Please explain that in client_state's blurb.

psarna commented:
v7:
 -  expanded tests
 - added a blurb about how updating per-role params works in client state
 - fixed typos/indentation

Current state of the follow-up list:

 * add propagating timeouts to all live sessions
 * add oltp/olap parameter
 * revive ALTER SESSION from previous iteration and make it available (for the sake of cqlsh and making it easier to test different values without having to have elevated privileges)
 * make internal statements use their own unique client state instead of a de-facto global state

psarna commented:
Actually, I think I'll also add one commit from `ALTER SESSION` right now - the one which fills system.clients table and allows peeking into the per-session parameters. That way, with multiple roles providing multiple timeouts, users would still be able to see what are the effective timeouts for given session.

avikivity commented on auth/standard_role_manager.cc:
```diff
@@ -132,6 +134,25 @@ const resource_set& standard_role_manager::protected_resources() const {
+                column_kind::regular_column);
+        schema_ptr table = b.build();
+        return mm.announce_column_family_update(table, false, {}).handle_exception([] (const std::exception_ptr&) {});
+    } catch (...) {
```
What happens if these run in parallel on two nodes?

I believe it should be fine. The new schema would be the same on both nodes, and receive the same digest.

@tgrabiec for verification.

avikivity commented on auth/service.cc:
```diff
@@ -403,14 +404,16 @@ future<> create_role(
     return ser.underlying_role_manager().create(name, config).then([&ser, name, &options] {
         if (!auth::any_authentication_options(options)) {
-            return make_ready_future<>();
+            return ser.underlying_role_manager().update_custom_options(name, options.role_options);
```
As I mentioned before, I don't believe we can hijack `CREATE ROLE ... WITH OPTIONS`. It is reserved for the authenticator. See [dse](https://docs.datastax.com/en/dse/6.7/cql/cql/cql_reference/cql_commands/cqlCreateRole.html#cqlCreateRole__cqlCreateRoleoptions-optionmap-axjdHcEh).

We'll need a new keyword, `SESSION OPTIONS = map`. A more SQL compatible/traditional format is to disaggregate the map: `WITH SESSION READ TIMEOUT = duration AND SESSION WRITE TIMEOUT = duration, but it's some amount of busy work with the parser. @kostja

avikivity commented on auth/service.cc:
```diff
@@ -403,14 +404,16 @@ future<> create_role(
                 ser.underlying_authenticator().supported_options()).then([&ser, name, &options] {
             return ser.underlying_authenticator().create(name, options);
+        }).then([&ser, name, &options] {
+            return ser.underlying_role_manager().update_custom_options(name, options.role_options);
```
One day we'll make this a transaction so you can't create a role part-way.

avikivity commented on cql3/statements/role-management-statements.cc:
```diff
@@ -68,11 +68,23 @@ using result_message = cql_transport::messages::result_message;
 using result_message_ptr = ::shared_ptr<result_message>;
 
 static auth::authentication_options extract_authentication_options(const cql3::role_options& options) {
+    static std::string_view allowed_role_options[] = {"read_timeout", "write_timeout"};
```
Ah, you overlap role-default-session options with custom authenticator options. I don't know if that's good. Maybe they should have a separate namespace.

avikivity replied:
(but maybe not)

psarna replied:
Yeah, I let them coexist here, assuming that it's unlikely that a custom authenticator option would be called `read_timeout` - especially that currently there are 0 authenticators which actually use these options at all. So it's a compromise between totally hijacking the options and introducing a special syntax.

avikivity commented on cql3/statements/role-management-statements.cc:
```diff
@@ -68,11 +68,23 @@ using result_message = cql_transport::messages::result_message;
-        authen_options.options = std::unordered_map<sstring, sstring>(options.options->begin(), options.options->end());
+        authen_options.options.emplace();
+        for (auto& entry : *options.options) {
+            if (std::find(std::begin(allowed_role_options), std::end(allowed_role_options), std::string_view(entry.first)) != std::end(allowed_role_options)) {
```
contains()

psarna replied:
`allowed_role_options` is an array now, due to having just 2 values, so I don't have a `contains` available. Rust traits would be handy here, but in C++ I think it doesn't get prettier than `std::find`.

avikivity commented on service/client_state.cc:
```diff
@@ -53,13 +53,64 @@
+    if (!_user || auth::is_anonymous(*_user)) {
+        return make_ready_future<>();
+    }
+    //FIXME: replace with a coroutine once they're widely accepted
```
```console
$ git grep -E 'co_await|co_return' | wc -l
44
```

psarna replied:
Yes, but
```bash
$ git grep -E 'co_await|co_return' | grep -c -v raft
0
```
, and raft has conditional compilation. I still expect frowns when pushing coroutines to unconditionally compiled code.

avikivity commented on cql3/statements/role-management-statements.cc:
```diff
@@ -118,6 +118,29 @@ future<> create_role_statement::grant_permissions_to_creator(const service::clie
+        }
+        if (duration.nanoseconds % 1'000'000 != 0) {
+            throw exceptions::invalid_request_exception("Timeout values must be expressed in millisecond granularity");
+        }
```
Let's also check they are strictly positive.

psarna replied:
I'd prefer non-negative. The reason being that while having a timeout value of 0 makes little sense on production clusters, it's great for deterministic tests.

avikivity commented on cql3/statements/role-management-statements.cc:
```diff
@@ -179,6 +180,8 @@ create_role_statement::execute(service::storage_proxy&,
         }).then([&state] () mutable {
             return state.get_client_state().update_per_role_params();
+        }).then([&state] () mutable {
+            return db::system_keyspace::update_per_session_params(state.get_client_state());
```
ALTER ROLE may affect a different role than the one we are running.

It should be left to the code that updates the timeouts already connected sessions.

psarna replied:
Yeah, it's pretty much lazy person's placeholder for easy testing. Ultimately, all sessions should simply get updated after these settings change.

dekimir replied:
OTOH, we could say the same thing about timeouts that we already say about permissions: the new settings only take hold after you log out and back in.

avikivity replied:
@psarna it also doesn't affect other nodes. Let's drop  this and do it correctly later.

avikivity replied:
> OTOH, we could say the same thing about timeouts that we already say about permissions: the new settings only take hold after you log out and back in.

That is actually pretty bad, since it means if you drop some permissions to restrict an application from doing bad things, you might not know you need to restart it.

Looks like we'll need a trigger mechanism to tell permissions and session watchers that something changed. I suppose the replica layer can broadcast a role_change event and cause caches to be flushed.

psarna replied:
Dropping it also makes the whole test suite invalid though :( So dropping this one is only feasible if we have some other mechanism for propagating the timeouts, because I'd rather not push this change without tests.

I can try adding proper propagation in this series as well, but it might end up being quite intrusive (literally, because chaining all client states in an intrusive list would be a way to go). Propagating to other shards would be quite easy, but as for other nodes, I guess we should observe schema changes and trigger an update when roles change?

psarna replied:
Alright, I thought about it more, and my decision is that I'll ensure proper propagation **within a node** in this series. It will enable testing and be a good first step. Then, we can perhaps touch inter-node propagation later?

psarna replied:
Sigh, `connection` class is already linked into an intrusive list and it makes it really easy to iterate over a snapshot of that list and update all the client states... the problem is that this intrusive list is from entirely another layer (it's a member of `cql_server`) and there's no way to reach it from role-updating statements. I would really prefer to avoid creating another intrusive list (over client states), but I don't see any clean alternatives.

psarna replied:
Maybe I'll just add a listener for a new `role_change` event.

avikivity replied:
@psarna a role change event is not enough, how do you trigger it on other nodes?

psarna replied:
In a follow-up, naturally! (: But, back to being serious, I plan to implement local propagation first and leave inter-node for later, because this pull request keeps growing. Propagating locally is not perfect, but it enables majority of the test scenarios. And, after such an event is implemented, propagating to other nodes will also get easier, since it would all be about generating an event in a proper place.

dekimir replied:
> > OTOH, we could say the same thing about timeouts that we already say about permissions: the new settings only take hold after you log out and back in.
> 
> That is actually pretty bad, since it means if you drop some permissions to restrict an application from doing bad things, you might not know you need to restart it.

I was talking more about role membership, which affects permissions indirectly.  I'm pretty sure that's not immediately propagated to all live sessions.  That's not ideal, but I doubt it's a surprise to any seasoned admin.

> Looks like we'll need a trigger mechanism to tell permissions and session watchers that something changed. I suppose the replica layer can broadcast a role_change event and cause caches to be flushed.

That works when roles are stored in Scylla but won't work when they're defined in an external source like LDAP.

dekimir replied:
> I was talking more about role membership, which affects permissions indirectly. I'm pretty sure that's not immediately propagated to all live sessions.

I should clarify this: the changes to `system_auth.roles` is propagated to all nodes fairly soon, but that table is not queried afresh on every statement executed in a session.  It's queried at login, and the result is cached for some time.

avikivity commented on cql3/statements/role-management-statements.cc:
```diff
@@ -491,6 +496,8 @@ grant_role_statement::execute(service::storage_proxy&, service::query_state& sta
     return as.underlying_role_manager().grant(_grantee, _role).then([&state] () mutable {
         return state.get_client_state().update_per_role_params();
+    }).then([&state] () mutable {
+            return db::system_keyspace::update_per_session_params(state.get_client_state());
```
ALTER ROLE psarna REDUCE INDENT;

avikivity commented:
Looks good. The only question is how to split role default session options from authenticator custom options:

1. no split, each consumer (authenticator/session) is responsible for picking unique names and filtering for them (this series)
2. add a namespace within the map ('CREATE ROLE foo WITH OPTIONS = { 'session$read_timeout' = 5s }`)
3. create a new options namespace ('CREATE ROLE foo WITH SESSION OPTIONS = { 'read_timeout' = 5s }`)
4. contribute CREATE SERVICE_LEVEL from enterprise (without workload prioritization, make this a service level option: `CREATE SERVICE_LEVEL analytics WITH READ TIMEOUT = 5s; ATTACH SERVICE_LEVEL analytics TO foo`

@eliransin / @kostja

tgrabiec commented:
On Tue, Dec 1, 2020 at 12:25 PM Avi Kivity <notifications@github.com> wrote:

> *@avikivity* commented on this pull request.
> ------------------------------
>
> In auth/standard_role_manager.cc
> <https://github.com/scylladb/scylla/pull/7617#discussion_r533335847>:
>
> > @@ -132,6 +134,25 @@ const resource_set& standard_role_manager::protected_resources() const {
>      return resources;
>  }
>
> +static future<> add_options_column_if_missing(database& db, ::service::migration_manager& mm) noexcept {
> +    try {
> +        auto schema = db.find_schema(sstring(meta::AUTH_KS), sstring(meta::roles_table::name));
> +        bytes options_name = to_bytes(meta::roles_table::options_col_name.data());
> +        if (schema->get_column_definition(options_name)) {
> +            return make_ready_future<>();
> +        }
> +        schema_builder b(schema);
> +        b.with_column(options_name,
> +                map_type_impl::get_instance(utf8_type, utf8_type, false),
> +                column_kind::regular_column);
> +        schema_ptr table = b.build();
> +        return mm.announce_column_family_update(table, false, {}).handle_exception([] (const std::exception_ptr&) {});
> +    } catch (...) {
>
> What happens if these run in parallel on two nodes?
>
> I believe it should be fine. The new schema would be the same on both
> nodes, and receive the same digest.
>
> @tgrabiec <https://github.com/tgrabiec> for verification.
>
The digest would be different on different nodes because timestamps would
be different. So it will result in schema exchange. For columns, if there
is no actual difference between nodes, schema merge will do nothing.

If it was for tables, it would drop the existing table.

avikivity commented:
> On Tue, Dec 1, 2020 at 12:25 PM Avi Kivity ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In auth/standard_role_manager.cc <[#7617 (comment)](https://github.com/scylladb/scylla/pull/7617#discussion_r533335847)>: > @@ -132,6 +134,25 @@ const resource_set& standard_role_manager::protected_resources() const { return resources; } +static future<> add_options_column_if_missing(database& db, ::service::migration_manager& mm) noexcept { + try { + auto schema = db.find_schema(sstring(meta::AUTH_KS), sstring(meta::roles_table::name)); + bytes options_name = to_bytes(meta::roles_table::options_col_name.data()); + if (schema->get_column_definition(options_name)) { + return make_ready_future<>(); + } + schema_builder b(schema); + b.with_column(options_name, + map_type_impl::get_instance(utf8_type, utf8_type, false), + column_kind::regular_column); + schema_ptr table = b.build(); + return mm.announce_column_family_update(table, false, {}).handle_exception([] (const std::exception_ptr&) {}); + } catch (...) { What happens if these run in parallel on two nodes? I believe it should be fine. The new schema would be the same on both nodes, and receive the same digest. @tgrabiec <https://github.com/tgrabiec> for verification.
> The digest would be different on different nodes because timestamps would be different. So it will result in schema exchange. For columns, if there is no actual difference between nodes, schema merge will do nothing. If it was for tables, it would drop the existing table.

Let's use a fixed timetstamp then, so that all the changes are synchronized.

psarna commented:
v8:
 * fixed more bad indents
 * made adding the `options` column in system_auth.roles with fixed timestamp
 * added propagating per-role values to all connections locally inside a node instead of only updating a single connection
 * added validation for durations to be non-negative

Pending review:
 * decide how to manage session-specific options (via coexisting with authenticator options (current state), adding a

Current state of the follow-up list:
 * namespace to session options, expanding the syntax (ALTER ROLLE WITH SESSION OPTIONS =), etc.
 * add oltp/olap parameter
 * revive ALTER SESSION from previous iteration and make it available (for the sake of cqlsh and making it easier to test different values without having to have elevated privileges)
 * make internal statements use their own unique client state instead of a de-facto global state
 * propagate role change events to other nodes, so that their sessions can also get updated without reconnecting

avikivity commented:
@eliransin @kostja please react to https://github.com/scylladb/scylla/pull/7617#issuecomment-736522242

kostja commented:
First, not once I asked to create an RFC for the feature, and not spread the discussion over a github issue.
Given there is no RFC, and the feature stopped using standard ANSI wordings, I lost interest.

One thing I don't understand regarding the switch from ALTER SESSION to CREATE ROLE ... OPTIONS is what happens if a user is granted multiple roles? Roles can be granted to roles. 

Other than that it seems reusing CREATE SERVICE LEVEL does not have ambiguity about what namespace to put the options into and what to do if two roles are merges.

psarna commented:
I keep updating this pull request's description in order to let people comment on it, and there's an entry in docs/ as part of this series, which can be reviewed and commented on. I prefer not to mix several mechanisms of review, so even though I'm not personally a github fan, I think that sending another description over e-mail, google docs or something else would just create multiple separate threads of discussion. We don't seem to have a single place for presenting RFC, so I just treat github as one of them. We can agree on a company-wide way of presenting them and I'll comply, but as far as I know it's not settled.

As for multiple roles, it depends on the parameter itself. In case of timeouts, the strictest possible combination of them is used (it's mentioned in the commit message, but not in the description, I'll fix it soon) - so for a `read_timeout`, a minimum of all values of all roles that somebody is a member of is effective.
For future parameters like oltp/olap - it doesn't make sense to mix the two of them. It's debatable how to solve this one - e.g. using one of the values and warn that multiple values are specified for this user, or disallowing to alter/grant a role with a conflicting value if one of the parent roles already have the value. But, since such parameter is not yet implemented, there's also no discussion of it yet.
I'll modify the description to have a paragraph on what happens with multiple roles, thanks.

psarna commented:
@avikivity should I proceed with CREATE SERVICE LEVEL then? I didn't dive into details yet, but I assume I should just borrow the syntax from enterprise, and only store the new parameters in the service level instances, without any workload prioritization code whatsoever.

avikivity commented:
I like ALTER SERVICE LEVEL. @eliransin is our for a few days, so he cannot comment. Please schedule a short meeting (today if possible) and present the options, while I'm anxious to get this in I don't want to make such decisions in an obscure pull request page.

psarna commented:
Well, I didn't manage to do the whole presentation today, but @vladzcloudius suggested to consider one more option - per query timeout. With that, we have even higher granularity than with per session, and I think that we can even minimize the overhead if we do this as a CQL extension, e.g.:
```cql
SELECT * FROM t TIMEOUT 50ms;
```
or
```cql
INSERT INTO t(a,b,c) VALUES (1,2,3) TIMEOUT 50ms;
```
By extending CQL we get integration with prepared statements for free, e.g.
```cql
SELECT * FROM t TIMEOUT ?;
```
or
```cql
INSERT INTO t(a,b,c) VALUES (?,?,?) TIMEOUT 50ms;
```

Sounds good to me.
/cc @avikivity

vladzcloudius commented:
On top of what Piotr wrote before I can add that this should probably also be implemented on a binary level too since drivers have their own timeout definitions which are per-query (they have per-session defaults too) and if these were propagated that would make all our DB level multiple timeout configurations obsolete.

psarna commented:
And as for oltp/olap - I suppose that can be implemented per-query as well. Given that it can be incorporated into a prepared statement, its overhead should be negligible.

avikivity commented:
> Well, I didn't manage to do the whole presentation today, 

I had an accomplice perform a diversion to disguise the fact I forgot to show up.

> but @vladzcloudius suggested to consider one more option - per query timeout. With that, we have even higher granularity than with per session, and I think that we can even minimize the overhead if we do this as a CQL extension, e.g.:
> 
> ```sql
> SELECT * FROM t TIMEOUT 50ms;
> ```
> 
> or
> 
> ```sql
> INSERT INTO t(a,b,c) VALUES (1,2,3) TIMEOUT 50ms;
> ```
> 
> By extending CQL we get integration with prepared statements for free, e.g.
> 
> ```sql
> SELECT * FROM t TIMEOUT ?;
> ```
> 
> or
> 
> ```sql
> INSERT INTO t(a,b,c) VALUES (?,?,?) TIMEOUT 50ms;
> ```
> 
> Sounds good to me.

Very good idea. But use the USING syntax as in USING TTL.

We'll still want session-scope (so we don't need to rewrite applications), but we an start with these overrides that are useful for testing.

We can also try USING DEADLINE, and so if the clocks are synchronized, we can account for time spent in queues before the server.

> /cc @avikivity

kostja commented:
* Avi Kivity <notifications@github.com> [20/12/09 21:14]:
> > ```sql
> > INSERT INTO t(a,b,c) VALUES (1,2,3) TIMEOUT 50ms;
> > ```

To not make parsing too complicated, it could be a good idea to
separate all statement-options with a single keyword.

<DML STATEMENT> <OPTION-START-KEYWORD> (<option-name>=<option_value>)+

OPTION-START-KEYWORD could be OPTION or USING  or something else
which is not possible.

Scylla CQL already has a place for USING TIMESTAMP, perhaps USING
TIMEOUT could be handled the same way from grammar point of view.


-- 
Konstantin Osipov, Moscow, Russia

psarna commented:
Going back to this discussion:

I think that the already merged per-query params + per-role/per-service-level params render per-session params deprecated. My rationale is that per-session params require the users to change their code anyway (by executing an ALTER SESSION statement before starting to do the actual work), so the users might just as well use per-query params, preferably combined with prepared statements. That way, each statement has its parameters explicitly specified and drivers are not forced to keep track of which session defined which parameters, because everything is transparently kept in the server-side prepared statement cache.
Conversely, if we want to be nice and allow users to use parameters like timeouts/oltp-olap without having to change their code at all, here's where per-role/per-service-level params come to play. The user code can be left as is, and all parameters can be separately specified for given roles and automatically propagated and applied to the whole cluster.

Opinions anyone?

avikivity commented:
I agree. Role-level is clearly needed for things like Spark. Session level is something intermediate between role-level and query-level, but it's not clear that it is so useful.

So I think we should move forward with service levels. We'll need the ability to react to changes for already-connected connections, but that can come later.

vladzcloudius commented:
> Going back to this discussion:
> 
> I think that the already merged per-query params + per-role/per-service-level params render per-session params deprecated. My rationale is that per-session params require the users to change their code anyway (by executing an ALTER SESSION statement before starting to do the actual work), so the users might just as well use per-query params, preferably combined with prepared statements. That way, each statement has its parameters explicitly specified and drivers are not forced to keep track of which session defined which parameters, because everything is transparently kept in the server-side prepared statement cache.
> Conversely, if we want to be nice and allow users to use parameters like timeouts/oltp-olap without having to change their code at all, here's where per-role/per-service-level params come to play. The user code can be left as is, and all parameters can be separately specified for given roles and automatically propagated and applied to the whole cluster.
> 
> Opinions anyone?

I think that this makes a lot of sense.

psarna commented:
Obsolete, per-service-level timeouts got in
